File: BestTimeToBuyAndSellStock.py
class Solution:
    # @param prices, a list of integer
    # @return an integer
    def maxProfit(self, prices):
        if not prices:
            return 0

        min_val, max_profit = prices[0], 0
        for i in xrange(1, len(prices)):
            min_val = min(min_val, prices[i])
            max_profit = max(max_profit, prices[i]-min_val)

        return max_profit

File: swift/BestTimeToBuyAndSellStock.swift
import Foundation


class Solution {
    func maxProfit(prices: [Int]) -> Int {
        if prices.isEmpty {
            return 0
        }
        
        var minVal = prices[0], maxProfit = 0
        
        for price in prices {
            minVal = min(minVal, price)
            maxProfit = max(maxProfit, price-minVal)
        }
        
        return maxProfit
    }
}